File: scripts/peach-ink-cup.ts
/**
 * Peach Ink Cup — 空色料杯 → 装满墨水 + Kwadron 风格背景
 *
 * 用法: npx tsx scripts/peach-ink-cup.ts
 *
 * 输出: output/peach_ink_cup/kwadron_pink.jpg
 *
 * 流程:
 * 1. 提示 Flux 生成一个装满 Peach 粉色墨水的色料杯
 * 2. 背景自动套 Kwadron 风格（深色渐变、硬光、工业美学）
 * 3. 生成 3 个配色变体（粉/绿/PMU透明）
 */
import fs from 'node:fs';
import path from 'node:path';
import https from 'node:https';
import { SocksProxyAgent } from 'socks-proxy-agent';
import 'dotenv/config';

const API_KEY = (process.env.REPLICATE_API_KEY || '').trim();
const MODEL = 'black-forest-labs/flux-1.1-pro-ultra';
const PROXY = process.env.PROXY || 'socks5://127.0.0.1:7890';
const agent = new SocksProxyAgent(PROXY);

const OUTPUT_DIR = 'F:/inkflow app/InkFlow_Project/Peach_AI_System/engine/output/peach_ink_cup';

function apiRequest(options: any, body?: string): Promise<any> {
  return new Promise((resolve, reject) => {
    const req = https.request({ ...options, agent }, res => {
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => {
        try { resolve(JSON.parse(data)); }
        catch { resolve(data); }
      });
    });
    req.on('error', reject);
    req.setTimeout(180000, () => { req.destroy(); reject(new Error('Timeout')); });
    if (body) req.write(body);
    req.end();
  });
}

async function download(url: string, dest: string): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    const file = fs.createWriteStream(dest);
    const req = https.get(url, { agent }, res => {
      if (res.statusCode && res.statusCode >= 300 && res.headers.location) {
        file.close();
        fs.unlinkSync(dest);
        https.get(res.headers.location, { agent }, r2 => {
          r2.pipe(file);
          file.on('finish', () => { file.close(); resolve(); });
        });
        return;
      }
      res.pipe(file);
      file.on('finish', () => { file.close(); resolve(); });
    });
    req.on('error', reject);
    req.setTimeout(60000);
  });
}

async function poll(getUrl: string, dest: string) {
  const u = new URL(getUrl);
  for (let i = 0; i < 60; i++) {
    await sleep(3000);
    const result: any = await apiRequest({
      hostname: u.hostname, path: u.pathname + u.search, method: 'GET',
      headers: { 'Authorization': `Bearer ${API_KEY}` },
    });
    console.log(`  [${i+1}] ${result.status}`);
    if (result.status === 'succeeded') {
      const url = Array.isArray(result.output) ? result.output[0] : result.output;
      await download(url, dest);
      return;
    }
    if (result.status === 'failed') {
      console.error('Failed:', JSON.stringify(result).slice(0, 400));
      return;
    }
  }
  console.error('Timeout');
}

async function generate(prompt: string, name: string) {
  console.log(`\n=== ${name} ===`);
  console.log('Prompt:', prompt.slice(0, 100) + '...');

  const postData = JSON.stringify({
    input: { prompt, aspect_ratio: '4:3', safety_tolerance: 5, raw: false, output_format: 'png' },
  });

  try {
    const result: any = await apiRequest({
      hostname: 'api.replicate.com',
      path: `/v1/models/${MODEL}/predictions`,
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${API_KEY}`,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData).toString(),
        'Prefer': 'wait',
      },
    }, postData);

    const dest = path.join(OUTPUT_DIR, name.replace(/\s+/g, '_') + '.png');

    if (result.status === 'succeeded' && result.output) {
      const url = Array.isArray(result.output) ? result.output[0] : result.output;
      await download(url, dest);
      console.log(`✓ Saved: ${dest}`);
    } else if (result.urls?.get) {
      console.log('Queued. Polling...');
      await poll(result.urls.get, dest);
    } else {
      console.log('Response:', JSON.stringify(result, null, 2).slice(0, 500));
    }
  } catch (err: any) {
    console.error('Error:', err.message);
  }
}

async function main() {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  // Kwadron 风格通用后缀
  const kwadronStyle = [
    'Professional product photography',
    'Hard light from upper left, sharp highlights on edges',
    'Pure black to dark blue gradient background, no texture',
    'High contrast cool tones, precision industrial aesthetic',
    'Shallow depth of field, studio quality, photorealistic',
    'No text, no watermark, no labels.',
  ].join(', ');

  const prompts = [
    {
      name: 'pink_ink_kwadron',
      prompt: [
        `A small silicone ink cup filled with vibrant peach pink liquid tattoo ink,`,
        `matte white silicone material, ink surface with soft reflection`,
        `${kwadronStyle}`,
      ].join(' '),
    },
    {
      name: 'green_ink_kwadron',
      prompt: [
        `A small silicone ink cup filled with vibrant mint green liquid tattoo ink,`,
        `matte white silicone material, ink surface with soft reflection`,
        `${kwadronStyle}`,
      ].join(' '),
    },
    {
      name: 'pmu_clear_ink_kwadron',
      prompt: [
        `A small silicone ink cup filled with translucent light pink PMU pigment ink,`,
        `matte white silicone material, ink appears delicate and sheer`,
        `${kwadronStyle}`,
      ].join(' '),
    },
    {
      name: 'black_ink_kwadron',
      prompt: [
        `A small silicone ink cup filled with deep black tattoo ink,`,
        `matte white silicone material, ink surface glossy and reflective`,
        `${kwadronStyle}`,
      ].join(' '),
    },
  ];

  for (const p of prompts) {
    await generate(p.prompt, p.name);
    await sleep(2000); // 避免太快
  }
}

const sleep = (ms: number) => new Promise(r => setTimeout(r, ms));
main().catch(e => { console.error('Fatal:', e?.message || e); process.exit(1); });

File: scripts/peach-ink-cup.py
#!/usr/bin/env python3
"""
Peach Ink Cup Generator — 空色料杯 → 装满墨水 + Kwadron 风格背景

用法: python3 scripts/peach-ink-cup.py
输出: output/peach_ink_cup/

生成 3 个 Peach 品牌配色变体（粉/绿/PMU透明）+ 1 个黑色
背景自动套 Kwadron 风格
"""
import os
import sys
import time
import json
import urllib.request
import urllib.error
from pathlib import Path

# ─── 配置 ───
API_KEY = os.getenv('REPLICATE_API_KEY', '').strip()
PROXY = os.getenv('PROXY', 'socks5://127.0.0.1:7890')
MODEL = 'black-forest-labs/flux-1.1-pro-ultra'
OUTPUT_DIR = Path(__file__).parent.parent / 'Peach_AI_System' / 'engine' / 'output' / 'peach_ink_cup'

def make_proxy():
    from urllib.request import ProxyHandler, build_opener
    # 去掉 socks5 用 http 代理，或用系统代理
    import socket
    socket.setdefaulttimeout(180)
    return build_opener()

def api_request(url, data=None, method='POST'):
    """调用 Replicate API"""
    import http.client
    from urllib.parse import urlparse
    
    parsed = urlparse(url)
    headers = {
        'Authorization': f'Bearer {API_KEY}',
        'Content-Type': 'application/json',
    }
    body = json.dumps(data).encode() if data else None
    if body:
        headers['Content-Length'] = str(len(body))
    
    print(f"  POST {url} ({len(body) if body else 0} bytes)")
    
    try:
        import http.client
        conn = http.client.HTTPSConnection('api.replicate.com', timeout=180)
        conn.request(method, url, body=body, headers=headers)
        resp = conn.getresponse()
        result = resp.read().decode()
        conn.close()
        return json.loads(result) if result else {}
    except Exception as e:
        print(f"  API Error: {e}")
        return None

def download_image(url, dest_path):
    """下载图片到本地"""
    import urllib.request
    try:
        req = urllib.request.Request(url, headers={
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'
        })
        urllib.request.urlretrieve(req, dest_path)
        size = os.path.getsize(dest_path)
        print(f"  ↓ Saved: {dest_path} ({size/1024:.0f} KB)")
        return True
    except Exception as e:
        print(f"  Download Error: {e}")
        return False

def generate_image(prompt, name, wait=True):
    """生成一张图片"""
    print(f"\n{'='*60}")
    print(f"Generating: {name}")
    print(f"Prompt: {prompt[:120]}...")
    
    # 创建prediction
    data = {
        'input': {
            'prompt': prompt,
            'aspect_ratio': '4:3',
            'safety_tolerance': 5,
            'raw': False,
            'output_format': 'png',
        }
    }
    
    result = api_request(
        f'/v1/models/{MODEL}/predictions',
        data=data
    )
    
    if not result:
        print("  ✗ No response from API")
        return False
    
    # 检查是否直接返回成功
    if result.get('status') == 'succeeded' and result.get('output'):
        url = result['output'][0] if isinstance(result['output'], list) else result['output']
        dest = OUTPUT_DIR / f"{name.replace(' ', '_')}.png"
        if download_image(url, str(dest)):
            return True
    
    # 需要轮询
    if result.get('urls', {}).get('get'):
        poll_url = result['urls']['get']
        print("  Queued. Polling...")
        
        for i in range(60):
            time.sleep(3)
            status_result = api_request(poll_url, method='GET')
            
            if not status_result:
                continue
                
            status = status_result.get('status', 'unknown')
            print(f"  [{i+1}] {status}")
            
            if status == 'succeeded':
                url = status_result['output'][0] if isinstance(status_result.get('output'), list) else status_result.get('output')
                dest = OUTPUT_DIR / f"{name.replace(' ', '_')}.png"
                if download_image(url, str(dest)):
                    return True
            elif status == 'failed':
                print(f"  ✗ Failed: {json.dumps(status_result).strip()[:200]}")
                return False
    
    print("  ✗ Timeout or error")
    return False

def main():
    # Kwadron 风格通用描述
    kwadron_style = (
        "Professional product photography, hard light from upper left creating sharp highlights, "
        "pure black to dark blue gradient background with no texture, "
        "high contrast cool tones with metallic edge glow, precision industrial aesthetic, "
        "shallow depth of field, studio quality, photorealistic, "
        "no text, no watermark, no labels."
    )
    
    prompts = [
        {
            'name': 'pink_ink_kwadron',
            'prompt': (
                f"A small white silicone ink cup filled with vibrant peach pink liquid tattoo ink, "
                f"ink surface with soft glossy reflection, matte silicone texture, "
                f"{kwadron_style}"
            )
        },
        {
            'name': 'green_ink_kwadron',
            'prompt': (
                f"A small white silicone ink cup filled with vibrant mint green liquid tattoo ink, "
                f"ink surface with soft glossy reflection, matte silicone texture, "
                f"{kwadron_style}"
            )
        },
        {
            'name': 'pmu_clear_ink_kwadron',
            'prompt': (
                f"A small white silicone ink cup filled with translucent light pink PMU pigment ink, "
                f"delicate sheer appearance, soft glossy ink surface, "
                f"{kwadron_style}"
            )
        },
        {
            'name': 'black_ink_kwadron',
            'prompt': (
                f"A small white silicone ink cup filled with deep black tattoo ink, "
                f"ink surface highly reflective and glossy, "
                f"{kwadron_style}"
            )
        },
    ]
    
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    
    for p in prompts:
        ok = generate_image(p['prompt'], p['name'])
        time.sleep(3)  # 避免太快
    
    print(f"\n{'='*60}")
    print(f"Done! Output directory: {OUTPUT_DIR}")
    print(f"Files: {list(OUTPUT_DIR.iterdir())}")

if __name__ == '__main__':
    main()
